### PR TITLE
Add custom matcher

### PR DESCRIPTION
**Pull request by AlexLisenkov**
- Added method to add custom matcher classes
- Removed else if statements
- [Optimized greedy array merge inside loop](https://github.com/kalessil/phpinspectionsea/blob/master/docs/performance.md#slow-array-function-used-in-loop)

### CODE REVIEW

File: src/Matcher.php [hunk 1]
```diff
@@ -3,9 +3,23 @@
 namespace ZxcvbnPhp;
 
 use ZxcvbnPhp\Matchers\Match;
+use ZxcvbnPhp\Matchers\MatchInterface;
 
 class Matcher
 {
+    private $defaultMatchers = [
```
Comment (jdreesen) [inline src/Matcher.php:10]: As this seems to be a constant array, why not make it a `const`?

File: src/Matcher.php [hunk 2]
```diff
@@ -78,16 +105,6 @@ public static function compareMatches(Match $a, Match $b)
...
-            Matchers\DateMatch::class,
-            Matchers\DictionaryMatch::class,
-            Matchers\ReverseDictionaryMatch::class,
-            Matchers\L33tMatch::class,
-            Matchers\RepeatMatch::class,
-            Matchers\SequenceMatch::class,
-            Matchers\SpatialMatch::class,
-            Matchers\YearMatch::class,
-        ];
+        return array_merge($this->defaultMatchers, $this->additionalMatchers);
```
Comment (jdreesen) [inline src/Matcher.php:108]: The `additionalMatchers` property contains string keys (the FQCN of the matcher) while the default matchers array has numeric indexes.

Should it be as follows to be consistent?

```suggestion
        return array_merge($this->defaultMatchers, array_values($this->additionalMatchers));
```

File: src/Matcher.php [hunk 3]
```diff
@@ -24,14 +38,27 @@ public function getMatches($password, array $userInputs = [])
...
 
+        $matches = array_merge([], ...$matches);
         self::usortStable($matches, [$this, 'compareMatches']);
+
         return $matches;
     }
 
+    public function addMatcher(string $className)
+    {
+        if (!in_array(MatchInterface::class, class_implements($className), true)) {
```
Comment (jdreesen) [inline src/Matcher.php:53]: This does the same but is easier to read, I think:

```suggestion
        if (!is_a($className, MatchInterface::class, true)) {
```

**Comment by bjeavons** [comment]
Hello and thank you @AlexLisenkov  and @jdreesen! Great feature suggestion and I appreciate the DX & perf changes. I'd like to have test coverage of this feature, at least for the exception. If that's something you don't have time for please let me know. Also cc @mkopinsky for any feedback. With test I'm inclined to commit.

**Comment by AlexLisenkov** [comment]
Hi, I've added tests that assert the exception and the return value of the new methods. Both classes are at 100% coverage now.

I've also upgraded phpunit to version 7 to match the minimum php 7.1 requirement.

**Comment by AlexLisenkov** [comment]
Hello maintainers, thank you for the quick feedback but after since there have been a bit of radio silence. Are we waiting on @mkopinsky feedback? Please let me know if there is anything I can do.

**Comment by bjeavons** [comment]
Merged to master. I'll cut 1.1.0 release soon. Thank you again @AlexLisenkov and @jdreesen!

**Pull request merged**